editoast/src/views/timetable/track_occupancy.rs

```diff
@@ -175,69 +209,105 @@ pub fn find_track_occupancy_for_operational_point(
     pathfinding: &PathfindingResult,
     train_schedule: &schemas::TrainSchedule,
 ) -> Vec<TrackOccupancy> {
-    // Extract pathfinding data
-    let (track_ranges, path_item_positions) = match pathfinding {
-        PathfindingResult::Success(pathfinding_result_success) => {
-            assert_eq!(
-                pathfinding_result_success.path_item_positions.len(),
-                train_schedule.path.len()
+    //Valid trains
+    if let (
+        simulation::Response::Success(SimulationResponseSuccess { final_output, .. }),
+        PathfindingResult::Success(pathfinding_result_success),
+    ) = (simulation, pathfinding)
+    {
+        let report_train = &final_output.report_train;
+        let path_projection =
+            PathProjection::new(&pathfinding_result_success.path.track_section_ranges);
+
+        if let Some(matching_index) =
+            find_matching_path_item(&train_schedule.path, operational_point_id, path_item_cache)
+        {
+            let schedule_item = train_schedule
+                .schedule
+                .iter()
+                .find(|schedule| schedule.at.0 == train_schedule.path[matching_index].id.0);
+            let arrival_time = report_train.path_item_times[matching_index] as i64;
+            let stop_duration = schedule_item
+                .and_then(|item| item.stop_for.clone())
+                .unwrap_or_default();
+
+            if let Some(track_section) = get_track_section_from_position(
+                pathfinding_result_success.path_item_positions[matching_index],
+                &path_projection,
+                operational_point_track_offsets,
+            ) {
+                return vec![TrackOccupancy {
+                    track_section,
+                    time_window: TimeWindow {
+                        time_begin: train_schedule.start_time
+                            + Duration::milliseconds(arrival_time),
+                        duration: stop_duration,
+                    },
+                }];
+            }
+        } else {
```

Comment:
You don't need an else condition when the matching if returns something.
The less indentation, the better.